### PR TITLE
build: support coverage testing and coveralls.io

### PR DESCRIPTION
**Pull request by grondo**
I was playing with adding code coverage support to `make check` and it actually went pretty smoothly, so I thought I'd submit the changes.

TL;DR --
 * new `--enable-code-coverage` option to `./configure` and `make check-code-coverage` target (`lcov(1)` package required)
 * Minor fixes submitted to get basic code coverage working
 * Support for [coveralls.io](https://coveralls.io/github/flux-framework/flux-core) code coverage tracking with each Travis-CI build

---
This PR adds `ax_code_coverage.m4` from [autoconf-archive](https://www.gnu.org/software/autoconf-archive/ax_code_coverage.html) and enables it by calling `AX_CODE_COVERAGE` in `configure.ac`.

This adds a new `configure` option `--enable-code-coverage`, which essentially turns on the flags:
```
-O0 -g -fprofile-arcs -ftest-coverage
```
and adds `-lgcov` to link lines.

I also added `@CODE_COVERAGE_LDFLAGS@` and `@CODE_COVERAGE_CFLAGS@`, which expand to the above when `--enable-code-coverage` is used, to `etc/Makefile.inc.in`. This is required to get 3rd party compiles using this Makefile to link with a flux-core built for profiling (e.g. tests under `t/t0006-build-basic.t`).

Finally, I added `@CODE_COVERAGE_RULES@` to `Makefile.am` which adds a convenience target  for
```
make check-code-coverage
``` 

which, runs `make check` then summarizes results into html using `lcov(1)` if available. (Actually not that useful, I'd be fine with dropping this one)

Unfortunately, all Makefiles also needed their `AM_CFLAGS` and `AM_LDFLAGS` updated to include `$(CODE_COVERAGE_CFLAGS)` and `$(CODE_COVERAGE_CPPFLAGS)`. There is probably an easier way to do that and preserve the ability to run `make` in a subdir, but I did not tackle that for now.

There was only one other fix required to get `make check` under coverage working. `flux lsattr` was hanging when flux-core was built with `--enable-code-coverage`. The failing build was against czmq-3.0.2 and I found that `attr_strcmp` for czmq > 3.0.1 was stil returning `bool`. I'm not sure why this only caused an issue when built with code coverage... perhaps the `-O0 -g`?

I also found from coverage testing that `hostlist.[ch]` isn't actually used by anything in `flux-core` anymore so these files were removed from liblsd.

This is the current result from running `make check-code-coverage && lcov --list flux*-coverage.info`
```
Reading tracefile flux-core-0.1.0-coverage.info
                                          |Lines       |Functions  |Branches    
Filename                                  |Rate     Num|Rate    Num|Rate     Num
================================================================================
[/home/travis/build/grondo/flux-core/src/]
bindings/lua/flux-lua.c                   |69.6%    855|65.2%    66|    -      0
bindings/lua/json-lua.c                   |84.0%     94|85.7%     7|    -      0
bindings/lua/kvs-lua.c                    |79.2%    125|86.7%    15|    -      0
bindings/lua/lalarm.c                     |87.5%     32| 100%     4|    -      0
bindings/lua/lutil.c                      |81.0%     42|80.0%     5|    -      0
bindings/lua/zmsg-lua.c                   |83.3%     84|80.0%    10|    -      0
bindings/python/flux/_barrier.c           |27.8%    126|20.0%    15|    -      0
bindings/python/flux/_core.c              | 7.5%   4992| 6.7%   421|    -      0
bindings/python/flux/_jsc.c               | 3.4%    380| 4.7%    43|    -      0
bindings/python/flux/_kvs.c               |10.6%   2103|12.0%   150|    -      0
bindings/python/flux/_mrpc.c              | 2.7%    476| 4.4%    45|    -      0
broker/attr.c                             |92.1%    140| 100%    17|    -      0
broker/boot_pmi.c                         |62.5%    160|61.9%    21|    -      0
broker/broker.c                           |77.3%   1199|83.8%    68|    -      0
broker/heartbeat.c                        |82.9%     70| 100%    12|    -      0
broker/hello.c                            |81.7%     93|80.0%    15|    -      0
broker/modservice.c                       |48.2%     85|50.0%     8|    -      0
broker/module.c                           |77.8%    334|81.8%    33|    -      0
broker/overlay.c                          |68.1%    373|82.2%    45|    -      0
broker/service.c                          |86.0%     43|83.3%     6|    -      0
broker/shutdown.c                         |97.3%     75| 100%    12|    -      0
broker/snoop.c                            |97.3%     37| 100%     8|    -      0
cmd/flux-comms-stats.c                    |44.3%     97|66.7%     3|    -      0
cmd/flux-comms.c                          |34.8%     66|50.0%     2|    -      0
cmd/flux-config.c                         |28.9%     90|33.3%     6|    -      0
cmd/flux-jstat.c                          |84.7%    111|88.9%     9|    -      0
cmd/flux-keygen.c                         |53.8%     26|50.0%     2|    -      0
cmd/flux-kvs.c                            |53.2%    314|69.6%    23|    -      0
cmd/flux-logger.c                         |67.5%     40|66.7%     3|    -      0
cmd/flux-module.c                         |66.7%    240|73.3%    15|    -      0
cmd/flux-mping.c                          |64.0%     50|50.0%     2|    -      0
cmd/flux-ping.c                           |81.4%    102|75.0%     4|    -      0
cmd/flux-start.c                          |70.6%    153|70.0%    10|    -      0
cmd/flux-zio.c                            |27.1%    361|21.1%    19|    -      0
cmd/flux.c                                |76.7%    270|89.5%    19|    -      0
cmd/forkzio.c                             |51.4%    138|77.8%     9|    -      0
common/libcompat/handle.c                 | 100%      8| 100%     3|    -      0
common/libcompat/info.c                   | 0.0%     20| 0.0%     4|    -      0
common/libcompat/reactor.c                |86.5%    163|95.0%    20|    -      0
common/libcompat/request.c                |79.7%     69| 100%     5|    -      0
common/libcompat/rpc.c                    |83.3%     12| 100%     1|    -      0
common/libflux/attr.c                     |92.0%    112|86.7%    15|    -      0
common/libflux/conf.c                     |79.5%    215|77.4%    31|    -      0
common/libflux/ev_flux.c                  |97.8%     45| 100%     6|    -      0
common/libflux/event.c                    |78.1%     32| 100%     2|    -      0
common/libflux/flog.c                     |98.6%     70|88.9%     9|    -      0
common/libflux/handle.c                   |82.1%    296|93.1%    29|    -      0
common/libflux/heartbeat.c                |84.6%     13| 100%     2|    -      0
common/libflux/info.c                     | 100%     12| 100%     3|    -      0
common/libflux/message.c                  |80.5%    599|95.9%    49|    -      0
common/libflux/module.c                   |91.4%    197| 100%    18|    -      0
common/libflux/panic.c                    | 0.0%     17| 0.0%     2|    -      0
common/libflux/reactor.c                  |85.6%    298|88.9%    36|    -      0
common/libflux/reduce.c                   |81.5%    146| 100%    10|    -      0
common/libflux/reparent.c                 |40.9%     22|50.0%     2|    -      0
common/libflux/request.c                  |78.1%     32| 100%     2|    -      0
common/libflux/response.c                 |75.9%     54| 100%     3|    -      0
common/libflux/rpc.c                      |83.7%    153|66.7%    15|    -      0
common/libflux/security.c                 |40.8%    348|55.6%    27|    -      0
common/libflux/tagpool.c                  |98.0%     50| 100%     6|    -      0
common/liblsd/cbuf.c                      |35.6%    776|37.8%    45|    -      0
common/liblsd/list.c                      |43.7%    316|42.9%    28|    -      0
common/libutil/base64.c                   |95.1%    142| 100%    10|    -      0
common/libutil/cleanup.c                  |97.3%     37| 100%     5|    -      0
common/libutil/coproc.c                   |87.0%     69| 100%     9|    -      0
common/libutil/env.c                      |69.2%     39|60.0%     5|    -      0
common/libutil/ev_zlist.c                 |94.7%     38| 100%     5|    -      0
common/libutil/ev_zmq.c                   |95.2%     42| 100%     5|    -      0
common/libutil/ev_zmq.h                   | 100%      5|    -     0|    -      0
common/libutil/ipaddr.c                   |72.7%     11| 100%     1|    -      0
common/libutil/jsonutil.c                 |49.6%    129|70.0%    20|    -      0
common/libutil/log.c                      |60.8%    125|80.0%    15|    -      0
common/libutil/monotime.c                 | 100%     13| 100%     3|    -      0
common/libutil/msglist.c                  |80.0%     80| 100%    12|    -      0
common/libutil/nodeset.c                  |80.4%    327|81.1%    37|    -      0
common/libutil/optparse.c                 |55.0%    331|46.2%    39|    -      0
common/libutil/readall.c                  | 0.0%     22| 0.0%     2|    -      0
common/libutil/sds.c                      |26.0%    453|34.2%    38|    -      0
common/libutil/sds.h                      |44.4%     63|75.0%     4|    -      0
common/libutil/setenvf.c                  | 0.0%      8| 0.0%     1|    -      0
common/libutil/shortjson.h                |92.2%    116| 100%    21|    -      0
common/libutil/tstat.c                    | 100%     34| 100%     8|    -      0
common/libutil/usa.c                      |76.8%     95|75.0%    16|    -      0
common/libutil/veb.c                      |91.2%    193| 100%    15|    -      0
common/libutil/veb_mach.c                 | 100%      3|    -     0|    -      0
common/libutil/vec.c                      |33.3%     57|37.5%     8|    -      0
common/libutil/xzmalloc.c                 |75.0%     32|83.3%     6|    -      0
connectors/local/local.c                  |88.9%    108| 100%     9|    -      0
connectors/loop/loop.c                    |89.3%     56| 100%     6|    -      0
connectors/shmem/shmem.c                  |78.9%    147|90.9%    11|    -      0
lib/libpmi/pmi.c                          |61.4%    228|44.1%    34|    -      0
modules/barrier/barrier.c                 |80.1%    156|80.0%    15|    -      0
modules/barrier/libbarrier.c              | 100%     28| 100%     2|    -      0
modules/connector-local/local.c           |71.8%    365|71.4%    28|    -      0
modules/kvs/conf.c                        |68.8%     48|75.0%     4|    -      0
modules/kvs/kvs.c                         |83.4%   1005|75.8%    66|    -      0
modules/kvs/libkvs.c                      |69.2%    757|72.5%    80|    -      0
modules/kvs/proto.c                       |75.9%    282| 100%    20|    -      0
modules/kvs/waitqueue.c                   |88.8%     89| 100%     9|    -      0
modules/libjsc/jstatctl.c                 |78.2%    527|87.9%    33|    -      0
modules/libkz/kz.c                        |62.2%    172|73.3%    15|    -      0
modules/libmrpc/mrpc.c                    |66.0%    191|68.8%    16|    -      0
modules/libsubprocess/subprocess.c        |84.9%    383|98.2%    55|    -      0
modules/libzio/zio.c                      |62.0%    463|62.1%    66|    -      0
modules/live/liblive.c                    | 0.0%     22| 0.0%     3|    -      0
modules/live/live.c                       |50.8%    524|44.7%    47|    -      0
modules/mecho/mecho.c                     |62.2%     37| 100%     2|    -      0
modules/modctl/libmodctl.c                |85.8%    148|85.7%     7|    -      0
modules/modctl/modctl.c                   |65.6%    131| 100%     5|    -      0
modules/modctl/proto.c                    |66.7%    147| 100%    15|    -      0
modules/pymod/py_mod.c                    | 0.0%     63| 0.0%     4|    -      0
modules/wreck/job.c                       |51.8%    139|71.4%     7|    -      0
modules/wreck/luastack.c                  |38.5%    182|42.9%    21|    -      0
modules/wreck/wrexec.c                    |48.7%    154|58.3%    12|    -      0
modules/wreck/wrexecd.c                   |73.9%    809|83.6%    73|    -      0
================================================================================
                                    Total:|51.0%  27876|56.6%  2475|    -      0
```

### CODE REVIEW

**Comment by grondo** [comment]
BTW, some other comments on the coveralls.io support:
 * It seemed strange to do coverage test for each compiler, so for now only the first build (which is typically GCC) does a coverage test for coveralls.
 * coverage doesn't play with `make distcheck` so the coverage build just does `make && make check-code-coverage`, skipping the distcheck. However, the second build still does distcheck.
 * lcov >= 0.10 seems to be required for coveralls to work, but only 0.9 is available from APT in Travis, so I had to include this package in `travis-dep-builder.sh` (FYI)
 * coveralls upload uses a private repo token set in the travis build environment for each repo (see for example https://travis-ci.org/flux-framework/flux-core/settings)

**Comment by grondo** [comment]
Well technically, coveralls should support running against PRs, but with the way the repo token is set, this does not currently work. I'll see if I can get that going since it might be useful, but for now you might be able to check out results here:

https://coveralls.io/github/grondo/flux-core

if coveralls allows public projects to be viewed.

**Comment by grondo** [comment]
Ok, I found that the COVERALLS_REPO_TOKEN isn't really necessary when running under Travis CI. I removed use of it and now coverage report for PR is working (click Show all checks above)
See coveralls.io results [here](https://coveralls.io/builds/3545389).

Still todo -- report source files that aren't in the coverage report (i.e. not run at all during `make check` so they have no .gcda/.gcno files)

**Comment by grondo** [comment]
Ok, I've just added a couple commits with a kludge to ensure all commands in src/cmd are at least run during a coverage test (runs all cmds with --help on `code-coverage-collect-hook`). I also had to fix `flux-jstat` which exited non-zero with `--help`.

I think the basic code coverage support is there now and this is ready to be merged.

**Comment by garlick** [comment]
This is pretty cool.  Merging!

**Comment by grondo** [comment]
Thanks! Check results [here](https://coveralls.io/github/flux-framework/flux-core)!

**Comment by lipari** [comment]
> I also found from coverage testing that hostlist.[ch] isn't actually used by anything in flux-core anymore so these files were removed from liblsd.
I'm guessing that broke the flux-sched build.  I was using hostset_within().  That's ok in that I've been meaning to replace that section of code using the resource.hwloc kvs instead.  So, don't worry about putting hostlist back into liblsd.

**Comment by grondo** [comment]
Ack, sorry about that, @lipari! 
A pretty amateur move to forget about flux-sched dependencies.
Another alternative is to pull in hostlist directly to flux-sched, but for just that one function it would not be worth it.
